### PR TITLE
Use DummyDataLoader instead of declaring new types for tests

### PR DESCRIPTION
**Pull request by shaahji**
## Use DummyDataLoader instead of declaring new types for tests

* Deprecate skip_dataloader because it ignores forced treating Dataset as Dataloader and we have no more use case for it.
* Updated DummyDataset
  * to generate random data instead of constant ones 
  * to accept a seed for randomness and a size
  * to avoid generating the data until explicitly requested
* Removed clones, yet slightly different, implementations of DummyDataset. Used primarily to tests.
* Renamed global create_dataloader to create_dummy_dataloader to be explicit. Used only in tests.

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: test/unit_test/passes/onnx/test_perf_tuning.py [hunk 1]
```diff
@@ -5,26 +5,33 @@
...
 from olive.evaluator.metric_result import flatten_metric_result
 from olive.evaluator.olive_evaluator import OliveEvaluator, OnnxEvaluator
 from olive.hardware.accelerator import DEFAULT_CPU_ACCELERATOR, DEFAULT_GPU_CUDA_ACCELERATOR, AcceleratorSpec, Device
 from olive.passes.olive_pass import create_pass_from_dict
 from olive.passes.onnx.perf_tuning import PERFTUNING_BASELINE, OrtPerfTuning, PerfTuningRunner, generate_test_name
 
 
+# TODO(shaahji): Remove this once perf_tuning pass supports DataConfig
+def _create_dummy_dataloader(data_dir, batch_size=1, **kwargs):
+    return DataLoader(DummyDataset([[1, 1]]), batch_size=batch_size)
```
Comment (jambayk) [inline test/unit_test/passes/onnx/test_perf_tuning.py:25]: doesn't the shape `[1,1]` already include the batch dimension. Is the data from this dataloader going to have shape `[1,1,1]` then?

File: test/unit_test/workflows/test_workfow_run.py
Comment (jambayk) [inline test/unit_test/workflows/test_workfow_run.py:1]: not related to the PR but just realized there is a typo in the name of this file

File: test/unit_test/passes/onnx/test_perf_tuning.py (hunk 1 above)
Comment (shaahji) [reply test/unit_test/passes/onnx/test_perf_tuning.py:25]: There is some ambiguity between the way tests are implementing and what perf tuning/metrics expect as the data.  The implementation of `DummyDataset` hasn't changed in what it used to return (except for using `rand()` instead of `ones()`). I have a follow up PR to make both tests and perf_tuning work consistently with batching input.

File: test/unit_test/utils.py [hunk 2]
```diff
@@ -193,8 +189,9 @@ def get_accuracy_metric(
         name="accuracy",
         type=MetricType.ACCURACY,
         sub_types=sub_types,
-        user_config=user_config or accuracy_metric_config,
+        user_config=user_config,
```
Comment (jambayk) [inline test/unit_test/utils.py:192]: looks like `accuracy_score_metric_config` is getting ignored?

Comment (jambayk) [reply test/unit_test/utils.py:192]: nevermind. I mixed the two configs.

File: test/unit_test/utils.py [hunk 3]
```diff
@@ -31,29 +33,9 @@ def forward(self, x):
...
-        self.data = torch.tensor(self.rng.random((size, 1)))
-        self.labels = torch.tensor(self.rng.random(1))
-
-    def __getitem__(self, idx):
-        return self.data[idx], self.labels[idx]
-
-    def __len__(self):
-        return self.size
+# TODO(shaahji): Remove this once perf_tuning pass supports DataConfig
+def create_dummy_dataloader(data_dir, batch_size=1, **kwargs):
```
Comment (jambayk) [inline test/unit_test/utils.py:37]: can you make the size of this dataset be `1`? Otherwise, the quantization_aware_training is taking a long time to run.

Comment (jambayk) [reply test/unit_test/utils.py:37]: the old dataloader also only had size 1

File: test/unit_test/passes/onnx/test_perf_tuning.py (hunk 1 above)
Comment (jambayk) [reply test/unit_test/passes/onnx/test_perf_tuning.py:25]: as long as the test passes, sounds good to me. Wasn't sure if it would pass since the shape `[1,1,1]` doesn't match the input shape of the model.

File: test/unit_test/utils.py (hunk 3 above)
Comment (shaahji) [reply test/unit_test/utils.py:37]: Done!

File: test/unit_test/workflows/test_workfow_run.py
Comment (shaahji) [reply test/unit_test/workflows/test_workfow_run.py:1]: Renamed!

File: test/unit_test/passes/onnx/test_perf_tuning.py (hunk 1 above)
Comment (shaahji) [reply test/unit_test/passes/onnx/test_perf_tuning.py:25]: All tests are passing. But, as you noticed, there are inconsistencies in inputs across tests and passes. Each test seems to be authored to work with specific data input, some accounting for batching and others don't.

**Review by jambayk — DISMISSED** [review]

File: olive/data/component/dataset.py [hunk 4]
```diff
@@ -115,20 +129,32 @@ def __getitem__(self, index):
...
-                dummy_inputs.append(torch.ones(shape, dtype=dtype))
-            dummy_inputs = tuple(dummy_inputs) if len(dummy_inputs) > 1 else dummy_inputs[0]
-        else:
+        # NOTE: torch.rand doesn't support all types, especially not int64 and bool
+        # So, always generate the data with dtype=float32 and convert it to expected input_type.
+        if self.input_names:
             dummy_inputs = {}
             for input_name, input_shape, input_type in zip(self.input_names, self.input_shapes, input_types):
-                dummy_inputs.update({input_name: torch.ones(input_shape, dtype=input_type)})
+                data = torch.rand(input_shape, generator=generator, dtype=torch.float32)
```
Comment (jambayk) [inline olive/data/component/dataset.py:143]: I think we still need an option or dataset that gives us constant ones for data. The dummy data for pytorch model also uses the dummy data container https://github.com/microsoft/Olive/blob/722749ecaba2117f36474b65b6efb38813ac6b3b/olive/model/handler/mixin/dummy_inputs.py#L28. This might cause issues with the `input_ids` and `attention_masks` being invalid or lead to unexpected behavior.

Comment (jambayk) [reply olive/data/component/dataset.py:143]: `torch.rand` and then cast to `int` leads to all zeros since the sampling is in `[0,1)`.

Comment (trajepl) [reply olive/data/component/dataset.py:143]: Also, dummy inputs may be used for model latency evaluation. Should we keep the inputs be consistent among model evaluations like torch, onnx, dlc and etc.?

File: olive/data/container/dummy_data_container.py [hunk 5]
```diff
@@ -18,17 +18,21 @@ class DummyDataContainer(DataContainer):
...
+                # TODO(shaahji): Remove this override once DataConfig transition is complete
+                params={ "batch_size": None }
+            )
         )
     """
 
     default_components_type: ClassVar[dict] = {
         DataComponentType.LOAD_DATASET.value: "dummy_dataset",
-        DataComponentType.DATALOADER.value: "skip_dataloader",
+        DataComponentType.DATALOADER.value: "default_dataloader",
```
Comment (trajepl) [inline olive/data/container/dummy_data_container.py:37]: `no_auto_batch_dataloader` takes the same effect with `default_dataloader` and `params={ "batch_size": None }`
image

File: olive/data/component/dataset.py (hunk 4 above)
Comment (shaahji) [reply olive/data/component/dataset.py:143]: Reverted to use `torch.ones` instead of `torch.rand`. A truly random dataset needs discussion and design. Fallback to what worked in the past for now.

File: olive/data/container/dummy_data_container.py (hunk 5 above)
Comment (shaahji) [reply olive/data/container/dummy_data_container.py:37]: Done!

File: test/unit_test/test_data_root.py [hunk 6]
```diff
@@ -53,6 +32,17 @@ def get_dataloader_config():
...
         },
+        "data_configs": [
+            {
+                "name": "test_data_config",
+                "type": "DummyDataContainer",
+                "load_dataset_config": {
+                    "type": "dummy_dataset",
+                    "params": {"data_dir": "data", "input_shapes": [1], "size": 1},
+                },
+                "dataloader_config": {"params": {"batch_size": 16}},
```
Comment (jambayk) [inline test/unit_test/test_data_root.py:43]: the dataloader component is `no_auto_batch_dataloader` so batch size is ignored

File: test/unit_test/passes/onnx/test_perf_tuning.py [hunk 7]
```diff
@@ -5,26 +5,33 @@
...
+
+
 @pytest.mark.parametrize(
     "config",
     [
         {"input_names": ["input"], "input_shapes": [[1, 1]]},
         {},
-        {"dataloader_func": create_dataloader},
-        {"dataloader_func": create_dataloader, "dataloader_func_kwargs": {"dummy_kwarg": 1}},
+        {"dataloader_func": _create_dummy_dataloader},
```
Comment (jambayk) [inline test/unit_test/passes/onnx/test_perf_tuning.py:33]: can this use the `create_dummy_dataloader` from `test.unit_tests.utils`?

File: test/unit_test/passes/pytorch/test_quantization_aware_training.py [hunk 8]
```diff
@@ -2,18 +2,26 @@
...
+from torch.utils.data import DataLoader
+
+from olive.data.component.dataset import DummyDataset
 from olive.hardware.accelerator import AcceleratorSpec
 from olive.passes.olive_pass import FullPassConfig, create_pass_from_dict
 from olive.passes.pytorch.quantization_aware_training import QuantizationAwareTraining
 
 
+# TODO(shaahji): Remove this once QuantizationAwareTraining pass supports DataConfig
+def _create_dummy_dataloader(data_dir, batch_size=1, size=1, **kwargs):
```
Comment (jambayk) [inline test/unit_test/passes/pytorch/test_quantization_aware_training.py:16]: same here

File: test/unit_test/utils.py [hunk 9]
```diff
@@ -162,23 +144,31 @@ def get_mock_openvino_model():
...
+        type="DummyDataContainer",
+        load_dataset_config=DataComponentConfig(
+            params={
+                "input_shapes": input_shapes,
+                "input_names": input_names,
+                "input_types": input_types,
+                "size": size,
+            }
+        ),
+        dataloader_config=DataComponentConfig(params={"batch_size": batch_size}),
```
Comment (jambayk) [inline test/unit_test/utils.py:159]: batch size is ignored here too. will probably have to update the callers to provide the correct shape?

Comment (shaahji) [reply test/unit_test/utils.py]: Fixed

File: test/unit_test/passes/pytorch/test_quantization_aware_training.py (hunk 8 above)
Comment (shaahji) [reply test/unit_test/passes/pytorch/test_quantization_aware_training.py]: This was done in response to your comment about time it takes for training. Required to force a size of 1 for dataset.

File: test/unit_test/passes/onnx/test_perf_tuning.py (hunk 7 above)
Comment (shaahji) [reply test/unit_test/passes/onnx/test_perf_tuning.py]: Done!

File: test/unit_test/test_data_root.py (hunk 6 above)
Comment (shaahji) [reply test/unit_test/test_data_root.py]: Fixed.

**Review by jambayk — APPROVED** [review]
